From 5382ed032a48ca6fefe8c3bca2c3c7c8edb8b63c Mon Sep 17 00:00:00 2001
From: Ken Russell <kbr@chromium.org>
Date: Thu, 10 Sep 2026 18:28:29 -0700
Subject: [PATCH] [M154] GL: apply host-twiddled TexSubImage2D workaround to
 all formats.

Original change's description:
> GL: apply host-twiddled TexSubImage2D workaround to all formats.
>
> The previous use_tex_sub_image_for_host_twiddled_npot_uploads
> workaround is needed for more texture formats, so simply apply it to
> all formats on affected GPUs, and rename the workaround to
> use_tex_sub_image_for_client_data_npot_uploads .
>
> Bug: chromium:559320837
> Change-Id: I588a6b0c664447e591bc2edb7a7682a103375ae2
> Reviewed-on: https://chromium-review.googlesource.com/c/angle/angle/+/8392118
> Auto-Submit: Kenneth Russell <kbr@chromium.org>
> Commit-Queue: Kenneth Russell <kbr@chromium.org>
> Reviewed-by: Zhenyao Mo <zmo@chromium.org>
> Reviewed-by: Geoff Lang <geofflang@chromium.org>

(cherry picked from commit a1962ac29d86eae65a17d20c3c181e8052acd90a)

Bug: chromium:561874871,chromium:559320837
Change-Id: I588a6b0c664447e591bc2edb7a7682a103375ae2
Reviewed-on: https://chromium-review.googlesource.com/c/angle/angle/+/8408594
Bot-Commit: rubber-stamper@appspot.gserviceaccount.com <rubber-stamper@appspot.gserviceaccount.com>
(cherry picked from commit 5382ed032a48ca6fefe8c3bca2c3c7c8edb8b63c)

---
 include/platform/autogen/FeaturesGL_autogen.h |  4 ++--
 include/platform/gl_features.json             |  4 ++--
 src/libANGLE/renderer/gl/TextureGL.cpp        | 13 ++-----------
 src/libANGLE/renderer/gl/renderergl_utils.cpp |  2 +-
 util/autogen/angle_features_autogen.cpp       |  2 +-
 util/autogen/angle_features_autogen.h         |  2 +-
 6 files changed, 9 insertions(+), 18 deletions(-)

diff --git a/third_party/angle/include/platform/autogen/FeaturesGL_autogen.h b/third_party/angle/include/platform/autogen/FeaturesGL_autogen.h
index a72ea6b3cf..da3ad5e255 100644
--- a/third_party/angle/include/platform/autogen/FeaturesGL_autogen.h
+++ b/third_party/angle/include/platform/autogen/FeaturesGL_autogen.h
@@ -272,8 +272,8 @@ struct FeaturesGL : FeatureSetBase
         &members,
     };
 
-    FeatureInfo useTexSubImageForHostTwiddledNpotUploads = {
-        "useTexSubImageForHostTwiddledNpotUploads",
+    FeatureInfo useTexSubImageForClientDataNpotUploads = {
+        "useTexSubImageForClientDataNpotUploads",
         FeatureCategory::OpenGLWorkarounds,
         &members,
     };
diff --git a/third_party/angle/include/platform/gl_features.json b/third_party/angle/include/platform/gl_features.json
index fd18975e48..2e6a242581 100644
--- a/third_party/angle/include/platform/gl_features.json
+++ b/third_party/angle/include/platform/gl_features.json
@@ -336,10 +336,10 @@
             "issue": "https://crbug.com/551573368"
         },
         {
-            "name": "use_tex_sub_image_for_host_twiddled_npot_uploads",
+            "name": "use_tex_sub_image_for_client_data_npot_uploads",
             "category": "Workarounds",
             "description": [
-                "For texture formats which would be twiddled on the host CPU, upload via TexSubImage2D instead of TexImage2D."
+                "For non-power-of-two textures, upload via TexSubImage2D instead of TexImage2D when uploading client-side data."
             ],
             "issue": "http://crbug.com/553172761"
         },
diff --git a/third_party/angle/src/libANGLE/renderer/gl/TextureGL.cpp b/third_party/angle/src/libANGLE/renderer/gl/TextureGL.cpp
index d7d875545d..fc5aec65ec 100644
--- a/third_party/angle/src/libANGLE/renderer/gl/TextureGL.cpp
+++ b/third_party/angle/src/libANGLE/renderer/gl/TextureGL.cpp
@@ -98,13 +98,6 @@ LevelInfoGL GetLevelInfo(const angle::FeaturesGL &features,
                        GetEmulatedAlphaChannel(features, originalInternalFormat));
 }
 
-bool IsHostTwiddledFormat(GLenum internalFormat, GLenum format, GLenum type)
-{
-    return (internalFormat == GL_RGB10_A2 && format == GL_RGBA &&
-            type == GL_UNSIGNED_INT_2_10_10_10_REV) ||
-           (internalFormat == GL_SRGB8_ALPHA8 && format == GL_RGBA && type == GL_UNSIGNED_BYTE);
-}
-
 gl::Texture::DirtyBits GetLevelWorkaroundDirtyBits()
 {
     gl::Texture::DirtyBits bits;
@@ -302,10 +295,8 @@ angle::Result TextureGL::setImageHelper(const gl::Context *context,
     if (nativegl::UseTexImage2D(getType()))
     {
         ASSERT(size.depth == 1);
-        if (features.useTexSubImageForHostTwiddledNpotUploads.enabled && pixels != nullptr &&
-            (!gl::isPow2(size.width) || !gl::isPow2(size.height)) &&
-            IsHostTwiddledFormat(texImageFormat.internalFormat, texImageFormat.format,
-                                 texImageFormat.type))
+        if (features.useTexSubImageForClientDataNpotUploads.enabled && pixels != nullptr &&
+            (!gl::isPow2(size.width) || !gl::isPow2(size.height)))
         {
             ANGLE_GL_TRY_ALWAYS_CHECK(
                 context, functions->texImage2D(
diff --git a/third_party/angle/src/libANGLE/renderer/gl/renderergl_utils.cpp b/third_party/angle/src/libANGLE/renderer/gl/renderergl_utils.cpp
index 87f428e45e..7b4573389c 100644
--- a/third_party/angle/src/libANGLE/renderer/gl/renderergl_utils.cpp
+++ b/third_party/angle/src/libANGLE/renderer/gl/renderergl_utils.cpp
@@ -2449,7 +2449,7 @@ void InitializeFeatures(const FunctionsGL *functions, angle::FeaturesGL *feature
     ANGLE_FEATURE_CONDITION(features, resetTexStorage2DBaseLevel, IsPowerVR(vendor));
 
     // crbug.com/553172761
-    ANGLE_FEATURE_CONDITION(features, useTexSubImageForHostTwiddledNpotUploads,
+    ANGLE_FEATURE_CONDITION(features, useTexSubImageForClientDataNpotUploads,
                             false /* IsPowerVR(vendor) */);
 
     ANGLE_FEATURE_CONDITION(features, adjustSrcDstRegionForBlitFramebuffer,
diff --git a/third_party/angle/util/autogen/angle_features_autogen.cpp b/third_party/angle/util/autogen/angle_features_autogen.cpp
index fc0d07eb90..b53627c3c1 100644
--- a/third_party/angle/util/autogen/angle_features_autogen.cpp
+++ b/third_party/angle/util/autogen/angle_features_autogen.cpp
@@ -512,7 +512,7 @@ constexpr PackedEnumMap<Feature, const char *> kFeatureNames = {{
     {Feature::UseStencilOpDynamicState, "useStencilOpDynamicState"},
     {Feature::UseStencilTestEnableDynamicState, "useStencilTestEnableDynamicState"},
     {Feature::UseSystemMemoryForConstantBuffers, "useSystemMemoryForConstantBuffers"},
-    {Feature::UseTexSubImageForHostTwiddledNpotUploads, "useTexSubImageForHostTwiddledNpotUploads"},
+    {Feature::UseTexSubImageForClientDataNpotUploads, "useTexSubImageForClientDataNpotUploads"},
     {Feature::UseUnusedBlocksWithStandardOrSharedLayout, "useUnusedBlocksWithStandardOrSharedLayout"},
     {Feature::UseVertexInputBindingStrideDynamicState, "useVertexInputBindingStrideDynamicState"},
     {Feature::UseVkEventForBufferBarrier, "useVkEventForBufferBarrier"},
diff --git a/third_party/angle/util/autogen/angle_features_autogen.h b/third_party/angle/util/autogen/angle_features_autogen.h
index 5ddb0103dc..30d1b30abb 100644
--- a/third_party/angle/util/autogen/angle_features_autogen.h
+++ b/third_party/angle/util/autogen/angle_features_autogen.h
@@ -512,7 +512,7 @@ enum class Feature
     UseStencilOpDynamicState,
     UseStencilTestEnableDynamicState,
     UseSystemMemoryForConstantBuffers,
-    UseTexSubImageForHostTwiddledNpotUploads,
+    UseTexSubImageForClientDataNpotUploads,
     UseUnusedBlocksWithStandardOrSharedLayout,
     UseVertexInputBindingStrideDynamicState,
     UseVkEventForBufferBarrier,
